app/src/main/java/net/gsantner/markor/format/ActionButtonBase.java

```diff
@@ -873,6 +880,37 @@ public static void moveLineSelectionBy1(final HighlightingEditor hlEditor, final
         }
     }
 
+    public static void duplicateLineSelection(final HighlightingEditor hlEditor) {
+        // Duplication is performed downwards, selection is moving alongside it.
+        final Editable text = hlEditor.getText();
+
+        final int[] sel = TextViewUtils.getSelection(hlEditor);
+        final int linesStart = TextViewUtils.getLineStart(text, sel[0]);
+        final int linesEnd = TextViewUtils.getLineEnd(text, sel[1]);
```

Comment:
look at TextViewUtils.getLineSelection and TextViewUtils.getSelectedLines

Reply:
Please see later comment on why I still think I need to use this existing approach.